### PR TITLE
feat(js/ai): added `maxTurns` option to generate options

### PR DESCRIPTION
**Pull request by pavelgj**
```ts
await ai.generate({
  prompt: 'call the tool',
  tools: ['testTool'],
  maxTurns: 17,
})
```

Issue: https://github.com/firebase/genkit/issues/1520

### CODE REVIEW

**Comment by mbleigh** [comment]
How about just "maxTurns"

File: js/ai/src/generate/action.ts [hunk 1]
```diff
@@ -240,7 +254,12 @@ async function generate(
     ] as MessageData[],
     tools: newTools,
   };
-  return await generateHelper(registry, nextRequest, middleware);
+  return await generateHelper(
+    registry,
+    nextRequest,
+    middleware,
+    toolCallIteration + 1
+  );
```
Comment (mbleigh) [inline js/ai/src/generate/action.ts:262]: Can we make sure the arguments are in the same order for generate and generateHeper? It's crossing my eyes that in one it's middleware, toolIteration and in the other it's toolIteration, middleware

File: js/ai/src/generate/action.ts [hunk 2]
```diff
@@ -79,8 +81,14 @@ export const GenerateUtilParamSchema = z.object({
 export async function generateHelper(
   registry: Registry,
   input: z.infer<typeof GenerateUtilParamSchema>,
-  middleware?: ModelMiddleware[]
+  middleware?: ModelMiddleware[],
+  toolCallIteration?: number
 ): Promise<GenerateResponseData> {
+  toolCallIteration = toolCallIteration ?? 0;
+  const maxIterations = input.maxTurns ?? 5;
```
Comment (mbleigh) [inline js/ai/src/generate/action.ts:88]: Maybe use the "turns" nomenclature everywhere?

File: js/ai/src/generate/action.ts [hunk 3]
```diff
@@ -79,8 +81,14 @@ export const GenerateUtilParamSchema = z.object({
...
+  if (toolCallIteration > maxIterations) {
+    throw new Error(`Exceeded maximum tool call iterations (${maxIterations})`);
+  }
```
Comment (mbleigh) [inline js/ai/src/generate/action.ts:91]: Should be GenkitError including the response with `ABORTED` status

Comment (pavelgj) [reply js/ai/src/generate/action.ts:91]: Done.

File: js/ai/src/generate/action.ts (hunk 1 above)
Comment (pavelgj) [reply js/ai/src/generate/action.ts]: done.

**Comment by pavelgj** [comment]
> How about just "maxTurns"

Done.

File: js/ai/src/generate/action.ts [hunk 4]
```diff
@@ -79,8 +81,20 @@ export const GenerateUtilParamSchema = z.object({
...
+  maxTurns?: number
 ): Promise<GenerateResponseData> {
+  maxTurns = maxTurns ?? 0;
+  const maxIterations = input.maxTurns ?? 5;
+  if (maxTurns > maxIterations) {
+    throw new GenkitError({
+      message: `Exceeded maximum tool call iterations (${maxIterations})`,
+      status: 'ABORTED',
+      detail: {
+        request: input,
```
Comment (mbleigh) [inline js/ai/src/generate/action.ts:94]: This should be a GenerationResponseError, I think, and include a response that includes all the stuff that was done until it bailed out.

Comment (pavelgj) [reply js/ai/src/generate/action.ts]: didn't know GenerationResponseError was a thing. :) Done.

**Review by mbleigh — APPROVED** [review]

**Pull request merged**